src/bitcoin-util.cpp

```diff
@@ -122,17 +119,19 @@ static int Grind(const std::vector<std::string>& args, std::string& strPrint)
     }
 
     uint32_t nBits = header.nBits;
-    std::atomic<bool> found{false};
+    std::atomic<std::optional<uint32_t>> found{};
 
     std::vector<std::thread> threads;
     int n_tasks = std::max(1u, std::thread::hardware_concurrency());
     for (int i = 0; i < n_tasks; ++i) {
-        threads.emplace_back( grind_task, nBits, std::ref(header), i, n_tasks, std::ref(found) );
+        threads.emplace_back(grind_task, nBits, header, i, n_tasks, std::ref(found));
```

Comment:
Maybe just pass in both a `std::atomic<bool>& found` and a `std::atomic<uint32_t>& proposed_nonce` with the threads doing:

```c++
if (hash <= target) { found = true; proposed_nonce = header.nNonce; return; }
```

and the parent doing
```c++
if (found) { header.nNonce = proposed_nonce }
else { error; }
```

Reply:
(gcc seems to treat `optional<atomic<int>>` and `atomic<bool> + atomic<int>` pretty similarly; clang seems to need to do calls to atomic_load/atomic_store with the optional which seems lame to me. mostly the code seems easier to read without the `.load()` and `.has_value()` stuff. But only very weak preference)

Reply:
Huh? Are you sure this will work. What if the thread is put to sleep after `found=true` and before setting the nonce. Then, the parent will succeed with `if(found)`, bug fail to read the nonce.

Reply:
Oh, I guess it requires the threads to join, which was the requirement before as well.

Reply:
Thanks, done